rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala

```diff
@@ -0,0 +1,317 @@
+          val prefixPath                   = commonPrefix(incomingPath, affixBytes)
+          val incomingHead :: incomingTail = incomingPath.drop(prefixPath.size)
+          val existingHead :: existingTail = affixBytes.drop(prefixPath.size)
+          val (dividedElems, dividedTopPtr) = divideSkip(
+            incomingTail,
+            incomingHead,
+            newLeaf,
+            existingTail,
+            existingHead,
+            existingPointer,
+            prefixPath
+          )
+          val updatedPointerBlock = pointerBlock.updated((toInt(common.last), dividedTopPtr) :: Nil)
+          for {
+            result <- rehash(common.init, elems, updatedPointerBlock, dividedElems)
+          } yield result
+
+        case TriePath(elems :+ (pastPb: PointerBlock), None, common) => // add to existing node
+          val key          = remainingPath.drop(common.size)
+          val (ptr, nodes) = skip(key, newLeaf)
+          val updatedPb    = pastPb.updated((toInt(common.last), ptr) :: Nil)
+          for {
+            result <- rehash(common.init, elems, updatedPb, nodes.toList)
+          } yield result
+
+        case TriePath(elems :+ Skip(affix, LeafPointer(_)), _, path) => // update value
+          rehash(path.dropRight(affix.size.toInt), elems, Skip(affix, newLeaf), Nil)
+        case _ => Sync[F].raiseError(MalformedTrieError)
+      }
+
+    private def rehash(
+        path: KeyPath,
+        rest: Vector[Trie],
+        lastSeen: Trie,
+        acc: List[Trie]
+    ): F[List[Trie]] = {
+      type Params = (KeyPath, Vector[Trie], Trie, List[Trie])
+      def go(params: Params): F[Either[Params, List[Trie]]] =
+        params match {
+          case (
+              remainingPath: KeyPath,
+              remainingPrefixNodes: Vector[Trie],
+              nextLastSeen: Trie,
+              currentAcc: List[Trie]
+              ) =>
+            (remainingPrefixNodes, nextLastSeen) match {
+              case (head :+ (pointerBlock: PointerBlock), other) =>
+                val ptr     = pointer(other)
+                val updated = pointerBlock.updated((toInt(remainingPath.last), ptr) :: Nil)
+                Applicative[F].pure(
+                  (remainingPath.init, head, updated, currentAcc :+ nextLastSeen).asLeft
+                )
+
+              case (head :+ Skip(affix, _), pb: PointerBlock) =>
+                Applicative[F]
+                  .pure(
+                    (
+                      remainingPath.dropRight(affix.size.toInt),
+                      head,
+                      Skip(affix, NodePointer(pb.hash)),
+                      currentAcc :+ nextLastSeen
+                    )
+                  )
+                  .map(_.asLeft)
+
+              case (Vector(), _) => Applicative[F].pure(currentAcc :+ nextLastSeen).map(_.asRight)
+              case _             => Sync[F].raiseError(MalformedTrieError)
+
+            }
+        }
+      FlatMap[F].tailRecM((path, rest, lastSeen, acc))(go)
```

Comment:
can you rewrite this `rehash` function as a fold instead of using recursivity?

Reply:
could you elaborate on the upside? please consider that the more I diverge from the original implementation the more it becomes hard to follow

Reply:
I understand. Nevermind.

Reply:
btw I agree that there is a fold there and I will try to refactor and see if its easier to read.